File: frontend/src/types/reasoning.ts
export interface ReasoningTrace {
  trace_id: string;
  workflow_id: string;
  iteration: number;
  phase: string;
  event_type: string;
  content: string;
  confidence: number | null;
  metadata: Record<string, unknown>;
  created_at: string;
}

export interface StructuredGoal {
  original_prompt: string;
  intent: string;
  constraints: string[];
  risk_level: string;
  expected_output_type: string;
  complexity_estimate: string;
  domain_hint: string | null;
  stop_conditions: string[];
}

File: backend/app/models/reasoning.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text, text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base


class ReasoningTrace(Base):
    __tablename__ = "reasoning_traces"

    trace_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()"),
    )
    workflow_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("workflows.workflow_id", ondelete="CASCADE"),
        nullable=False, index=True,
    )
    iteration: Mapped[int] = mapped_column(Integer, nullable=False, server_default=text("0"))
    phase: Mapped[str] = mapped_column(String(50), nullable=False)
    event_type: Mapped[str] = mapped_column(String(50), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    confidence: Mapped[float | None] = mapped_column(Float, nullable=True)
    metadata_: Mapped[dict] = mapped_column(
        "metadata", JSONB, nullable=False, server_default=text("'{}'::jsonb"),
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=text("now()"), nullable=False,
    )

    def __repr__(self) -> str:
        return f"<ReasoningTrace(phase={self.phase}, event={self.event_type})>"
